modules/infoserv/main.c

```diff
@@ -326,159 +289,250 @@ is_cmd_post(struct sourceinfo *si, int parc, char *parv[])
+	if (!to_id || to_id > list->count)
+	{
+		command_fail(si, fault_nosuch_target, _("The target position does not exist."));
+		return;
+	}
 
-			mowgli_node_delete(n, &operlogon_info);
+	if (from_id == to_id)
+	{
+		command_fail(si, fault_nochange, _("You must specify two different positions."));
+		return;
+	}
 
-			strshare_unref(o->nick);
-			sfree(o->subject);
-			sfree(o->story);
-			sfree(o);
+	mowgli_node_t *from_node = NULL, *to_node = NULL, *n;
+	struct logoninfo *entry = NULL;
 
-			command_success_nodata(si, _("Deleted entry %u from oper logon info."), id);
-			return;
+	unsigned int pos = 0;
+	MOWGLI_ITER_FOREACH(n, list->head)
+	{
+		// 1-based indexing, reflecting what we display in LIST output
+		pos++;
+
+		if (!from_node && pos == from_id)
+		{
+			// We'll be removing this node from the list and adding it before the
+			// item currently at the target position. We need to skip this node
+			// in the count in case the target position is after it, however, as
+			// we're going to be removing it.
+			pos--;
+			from_node = n;
+			entry = n->data;
 		}
+
+		if (!to_node && pos == to_id)
+			to_node = n;
+
+		if (from_node && to_node)
+			break;
+	}
+
+	// this should never happen
+	// note that to_node can be NULL if we're asked to move an item to the end of the list;
+	// mowgli_node_insert_before will do the right thing in that case
+	if (!from_node || from_node == to_node)
+	{
+		if (!from_node)
+			slog(LG_ERROR, "%s: failed to find node %u in list (BUG)", MOWGLI_FUNC_NAME, from_id);
+		else
+			slog(LG_ERROR, "%s: trying to move node %u to position %u but found the same node (BUG)", MOWGLI_FUNC_NAME, from_id, to_id);
+
+		command_fail(si, fault_internalerror, _("There was an error modifying the list. This is a bug in services."));
+		return;
 	}
 
-	command_fail(si, fault_nosuch_target, _("Entry \2%u\2 not found in oper logon info."), id);
+	mowgli_node_delete(from_node, list);
+	mowgli_node_add_before(entry, from_node, list, to_node);
+
+	if (oper)
+		command_success_nodata(si, _("Oper logon info entry \2%u\2 moved to position \2%u\2."), from_id, to_id);
+	else
+		command_success_nodata(si, _("Logon info entry \2%u\2 moved to position \2%u\2."), from_id, to_id);
+	logcommand(si, CMDLOG_ADMIN, "INFO:%s: \2%u\2 -> \2%u\2: \2%s\2, \2%s\2", oper ? "OMOVE" : "MOVE", from_id, to_id, entry->subject, entry->story);
 	return;
 }
 
 static void
-is_cmd_list(struct sourceinfo *si, int parc, char *parv[])
+is_cmd_move(struct sourceinfo *si, int parc, char *parv[])
+{
+	is_move_impl(si, parc, parv, false);
+}
+
+static void
+is_cmd_omove(struct sourceinfo *si, int parc, char *parv[])
+{
+	is_move_impl(si, parc, parv, true);
+}
+
+static void
+is_list_impl(struct sourceinfo *si, int parc, char *parv[], bool oper)
 {
-	struct logoninfo *l;
 	mowgli_node_t *n;
-	struct tm *tm;
-	char dBuf[BUFSIZE];
 	unsigned int x = 0;
+	mowgli_list_t list = oper ? operlogon_info : logon_info;
```

Comment:
This makes the compiler emit a pointless memory copy for struct assignment; take its address and use a pointer instead.